@namespace BlazorApp.Components
@using Confluent.Kafka
@inject State State;
@inject KafkaSettings KafkaSettings

<div class="topic">
    <h3>Topic <TopicSelector @bind-Topic="Name" /> <button @onclick="Sync">Sync</button></h3>
    <div class="partitions">
        @foreach (var partition in State.Partitions) {
            <Partition Data="@partition" />
        }
    </div>
</div>

@code {
    public string? Name { get; set; }

    protected override void OnInitialized() {
        base.OnInitialized();

        State.StateChanged += OnStateChanged;
    }

    private void OnStateChanged(object? sender, EventArgs args) {
        InvokeAsync(StateHasChanged);
    }

    private void Sync() {
        State.Clear();
        State.IsBrokerReady = false;

        var partitionSyncConsumerConfig = new ConsumerConfig
        {
            BootstrapServers = KafkaSettings.BootstrapServers,
            SaslUsername = KafkaSettings.SaslUsername,
            SaslPassword = KafkaSettings.SaslPassword,
            SecurityProtocol = SecurityProtocol.SaslSsl,
            SaslMechanism = SaslMechanism.Plain,
            GroupId = "sync-partitions",
            AutoOffsetReset = AutoOffsetReset.Latest,
            EnableAutoCommit = false,
        };

        var cancellationTokenSource = new CancellationTokenSource();
        using (var consumer = new ConsumerBuilder<Ignore, string>(partitionSyncConsumerConfig)
        .SetPartitionsAssignedHandler((consumer, partitions) =>
        {
            foreach (var partition in partitions)
            {
                State.AddPartition(partition.Partition.Value);
            }
            State.IsBrokerReady = true;
            InvokeAsync(StateHasChanged);
            cancellationTokenSource.Cancel();
        }).Build())
        {
            consumer.Subscribe(Name);
            try
            {
                consumer.Consume(cancellationTokenSource.Token);
            }
            catch (OperationCanceledException)
            {
                consumer.Close();
            }
            catch (ConsumeException)
            {
                consumer.Close();
            }
        }
    }
}